mgr/cephadm: skip ok-to-stop for mds in upgrade if not enough mds daemons

adk3798 opened the pull request:
Fixes: https://tracker.ceph.com/issues/50817
    
Signed-off-by: Adam King <adking@redhat.com>

<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

mgfritch reviewed (commented):
I'm not sure this works since a `.` char is valid within an fs_name

for example:
```
$ ceph fs volume create a.b.c
```

adk3798 commented:
> I'm not sure this works since a `.` char is valid within an fs_name
> 
> for example:
> 
> ```
> $ ceph fs volume create a.b.c
> ```

changed this to use `.rsplit('.', 2)[0]` to get fs name from daemon id. I think that should work as long as the mds daemon id follows the form `<fs-name>.<hostname>.<random-chars>` that we're using. Also changed a couple of the cases in the unit test to use a fs name with a "." in it.

adk3798 commented:
> I'm not sure this works since a `.` char is valid within an fs_name
> 
> for example:
> 
> ```
> $ ceph fs volume create a.b.c
> ```

This now uses `mds_daemon.service_name().split('.', 1)[1]` to match the mds daemon to its fs. Since the service name we store is always `mds.<fs-name>` I think it should work regardless of the number of "." characters in the fs name or hostname of the host the daemon is on. Now also using Cephadm's HostCache to get the number of daemons rather than the mds metadata since it's too difficult to match the mds daemons in the metadata to the correct fs.

mgfritch commented:
jenkins test make check

sebastian-philipp commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -433,6 +433,31 @@ def _enough_mons_for_ok_to_stop(self) -> bool:
+                return True
+            return False
+
+        raise OrchestratorError(f'mds daemon {mds_daemon.name()} does not belong to any fs')
```
This can happen, if you call `ceph orch apply mds fs_name` instead of `ceph fs volume create ...`. Would it make sense to succeed here?

adk3798 replied:
I just checked and it looks like mds daemons without a fs just pass ok-to-stop every time regardless of how many of them there are in the service so yes, we should be good to succeed here.

sebastian-philipp commented:
https://pulpito.ceph.com/swagner-2021-06-15_14:12:12-rados:cephadm-wip-swagner-testing-2021-06-15-1350-distro-basic-smithi/
